File: backend/src/main/java/com/inso/sila/entity/StudioActivity.java
package com.inso.sila.entity;

import com.fasterxml.jackson.annotation.JsonBackReference;
import com.inso.sila.enums.ActivityType;
import jakarta.persistence.CascadeType;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.FetchType;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToMany;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.OneToOne;
import jakarta.persistence.Table;
import lombok.Data;
import lombok.Builder;
import lombok.AllArgsConstructor;
import lombok.EqualsAndHashCode;
import lombok.NoArgsConstructor;
import lombok.Getter;
import lombok.Setter;

import java.time.LocalDateTime;
import java.util.List;

@Data
@Entity
@Builder
@Getter
@Setter
@AllArgsConstructor
@NoArgsConstructor
@Table(name = "studio_activity")
@EqualsAndHashCode
public class StudioActivity {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "studio_activity_id")
    private Long studioActivityId;

    @Column(nullable = false)
    private String name;

    @Column(columnDefinition = "TEXT")
    private String description;

    @OneToOne(cascade = CascadeType.ALL)
    @JoinColumn(name = "profile_image_id")
    private ProfileImage profileImage;

    @Column(nullable = false, name = "date_time")
    private LocalDateTime dateTime;

    @Column(nullable = false)
    private float duration;

    @Column(nullable = false)
    private float price;

    @Column(nullable = false)
    private ActivityType type;

    @Column
    private int capacity;

    @ManyToMany(fetch = FetchType.EAGER, mappedBy = "studioActivities", cascade = {CascadeType.DETACH, CascadeType.MERGE, CascadeType.PERSIST, CascadeType.REFRESH})
    @JsonBackReference
    private List<ApplicationUser> applicationUsers;

    @OneToOne
    private Membership membership;

    @OneToOne(fetch = FetchType.LAZY)
    private StudioActivityPreferences preferences;

    @ManyToOne(fetch = FetchType.EAGER)
    @JoinColumn(name = "instructor_id")
    private Instructor instructor;

}

File: frontend/src/app/dtos/studioActivity.ts
import {Membership} from "./membership";
import {UserCreateDto} from "./user";
import {ProfileImageDto} from "./image";
import {Instructor} from "./instructor";
import {SkillLevel} from "../enums/skillLevel";

export class StudioActivity {
  studioActivityId: number;
  profileImage: ProfileImageDto;
  name: string;
  description: string;
  dateTime: Date;
  duration: number;
  price: number;
  type: string;
  membership?: Membership;
  capacity?: number;
  applicationUsers?: UserCreateDto[];
  instructor?:Instructor;
}

export interface StudioActivityListDto {
  studioActivityId: number;
  name: string;
  profileImage: ProfileImageDto;
  description: string;
  dateTime: Date;
  duration: number;
  price: number;
}

export class StudioActivityCreateDto {
  name: string;
  profileImageFile: File;
  description: string;
  dateTime: string;
  duration: number;
  price: number;
  type: string;
  studioId: string;
  capacity: string;
  skillLevel: SkillLevel;
  equipment: boolean;
}


export interface SearchActivitiesDto {
  activityType: string;
  pageIndex: number;
  pageSize: number;

}

export interface StudioActivityTypeSearchResponseDto {
  studioActivityId: number;
  name: string;
  description: string;
  dateTime: Date;
  duration: number;
  price: number;
  profileImagePath: string;
}
